pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -144,23 +147,70 @@ type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
 // the output of the stringSliceFunc to be sorted, as results of LabelValues
 // are not sorted.
 //
-// TODO: Consider running stringSliceFunc calls concurrently
+
+type stringSliceFuncResult struct {
+	Result   []string
+	Warnings storage.Warnings
+}
+
 func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+
+	g, _ := errgroup.WithContext(m.ctx)
+
 	var warnings storage.Warnings
 	resultMap := make(map[string]struct{})
-	for pos, tenantID := range m.tenantIDs {
-		result, resultWarnings, err := f(m.queriers[pos])
-		if err != nil {
-			return nil, nil, err
-		}
-		for _, e := range result {
-			resultMap[e] = struct{}{}
-		}
-		for _, w := range resultWarnings {
-			warnings = append(warnings, fmt.Errorf("error querying tenant id %s: %w", tenantID, w))
+	resultCh := make(chan *stringSliceFuncResult)
+	resultDone := make(chan struct{})
+
+	go func() {
+		defer func() {
+			close(resultDone)
+		}()
+		for result := range resultCh {
+			for _, e := range result.Result {
+				resultMap[e] = struct{}{}
+			}
+			warnings = append(warnings, result.Warnings...)
 		}
+	}()
+
+	for pos := range m.tenantIDs {
+		querier := m.queriers[pos]
+		tenantID := m.tenantIDs[pos]
+		g.Go(func() error {
+			result, resultWarnings, err := f(querier)
+			if err != nil {
+				return errors.Wrapf(err, `error querying {%s="%s"}`, defaultTenantLabel, tenantID)
+			}
+
+			var warnings = make(storage.Warnings, len(resultWarnings))
```

Comment:
Nit: we can reuse `resultWarnings` and put wrapped warnings back to it.

Reply:
I have looked into this a fair bit now, but I am a bit stuck. When I make the changes as @pstibrany suggested I get a race condition (see https://github.com/cortexproject/cortex/commit/b6eee271a0a9cc65a17fc1d6ad2a6de18ace2cb3):

```
                Error Trace:    merge_queryable_test.go:447
                                                        merge_queryable_test.go:432
                Error:          elements differ

                                extra elements in list A:
                                ([]interface {}) (len=2) {
                                 (string) (len=58) "warning querying {__tenant_id__=\"team-b\"}: don't like them",
                                 (string) (len=56) "warning querying {__tenant_id__=\"team-c\"}: out of office"
                                }


                                extra elements in list B:
                                ([]interface {}) (len=2) {
                                 (string) (len=99) "warning querying {__tenant_id__=\"team-c\"}: warning querying {__tenant_id__=\"team-c\"}: out of office",
                                 (string) (len=101) "warning querying {__tenant_id__=\"team-b\"}: warning querying {__tenant_id__=\"team-b\"}: don't like them"
                                }


                                listA:
                                ([]string) (len=2) {
                                 (string) (len=58) "warning querying {__tenant_id__=\"team-b\"}: don't like them",
                                 (string) (len=56) "warning querying {__tenant_id__=\"team-c\"}: out of office"
                                }


                                listB:
                                ([]string) (len=2) {
                                 (string) (len=99) "warning querying {__tenant_id__=\"team-c\"}: warning querying {__tenant_id__=\"team-c\"}: out of office",
                                 (string) (len=101) "warning querying {__tenant_id__=\"team-b\"}: warning querying {__tenant_id__=\"team-b\"}: don't like them"
                                }
                Test:           TestMergeQueryable/three_tenants_with_some_return_storage_warnings
```